main/src/library/List.flix

```diff
@@ -0,0 +1,579 @@
+    /// That is, scanLeft(f, s, xs) = s :: scanLeftHelper(f, s, xs).
+    ///
+    def scanLeftHelper[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(s, x) :: scanLeftHelper(f, f(s, x), rs)
+    }
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going right to left.
+    /// That is, the result is of the form: `... f(xn-1, f(xn, s)) :: f(xn, s) :: s`.
+    ///
+    def scanRight[a,b](f: (a, b) -> b, s: b, xs: List[a]): List[b] = match xs with {
+        case Nil => s :: Nil
+        case x :: rs => match scanRight(f, s, rs) with {
+            case y :: ys => f(x, y) :: y :: ys
+        }
+    }
+
+    //
+    // ## List Transformation
+    //
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs`.
+    /// That is, the result is of the form: `f(x1) :: f(x2) :: ...`.
+    ///
+    def map[a,b](f: a -> b, xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x) :: map(f, rs)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index.
+    /// That is, the result is of the form: `f(x1, 0) :: f(x2, 1) :: ...`.
+    ///
+    def mapWithIndex[a,b](f: (a, Int32) -> b, xs: List[a]): List[b] = mapWithIndexHelper(f, xs, 0)
+
+    ///
+    /// Helper function for `mapWithIndex`.
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index + `i`.
+    /// That is, the result is of the form: `f(x1, i) :: f(x2, i+1) :: ...`.
+    /// mapWithIndex(f, xs) = mapWithIndexHelper(f, xs, 0).
+    ///
+    def mapWithIndexHelper[a,b](f: (a, Int32) -> b, xs: List[a], i: Int32): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i+1)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    ///
+    def flatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => append(f(x), flatMap(f, rs))
+    }
+
+    ///
+    /// Returns the reverse of `xs`.
+    ///
+    def reverse[a](xs: List[a]): List[a] = reverseHelper(xs, Nil)
+
+    ///
+    /// Helper function for `reverse`.
+    /// Returns `acc` appended to the reverse of `xs`.
+    /// reverse(xs) = reverseHelper(xs, Nil).
+    ///
+    def reverseHelper[a](xs: List[a], acc: List[a]): List[a] = match xs with {
+        case Nil => acc
+        case x :: rs => reverseHelper(rs, x :: acc)
+    }
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the left.
+    /// That is, returns a new list where the first `n mod length(xs)` elements in `xs`
+    /// are the last `n mod length(xs)` elements of the new list.
+    ///
+    def rotateLeft[a](n: Int32, xs: List[a]): List[a] =
```

Comment:
`rotateLeft` and `rotateRight` are now valid for any `n`.  The distance rotated is the positive value of `n mod length(xs)`.